momento-sdk/src/main/java/momento/sdk/batchutils/MomentoBatchUtils.java

```diff
@@ -0,0 +1,188 @@
+     */
+    public MomentoBatchUtilsBuilder(final CacheClient cacheClient) {
+      this.cacheClient = cacheClient;
+    }
+
+    /**
+     * Sets the maximum number of concurrent requests allowed in a batch operation. This limit is
+     * important for controlling the load on both the client and server side. Increasing this number
+     * can lead to a higher network and server stress. It is advised to keep this number reasonable
+     * and not exceed a certain threshold (e.g., 100), as higher values may significantly impact the
+     * application's performance and stability, particularly under constrained network conditions or
+     * limited server resources.
+     *
+     * @param maxConcurrentRequests The maximum number of concurrent requests. Caution should be
+     *     exercised if setting this value above 100.
+     * @return The builder instance for chaining.
+     */
+    public MomentoBatchUtilsBuilder withMaxConcurrentRequests(int maxConcurrentRequests) {
+      this.maxConcurrentRequests = maxConcurrentRequests;
+      return this;
+    }
+
+    /**
+     * Sets the maximum timeout for each individual asynchronous request that is sent to Momento. In
+     * theory these requests should never take any disk or network time as it's simply submitting an
+     * async request, so the default 10 second timeout is a more than conservative amount for the
+     * same.
+     *
+     * @param requestTimeoutSeconds The maximum timeout for each individual request to Momento.
+     * @return The builder instance for chaining.
+     */
+    public MomentoBatchUtilsBuilder withRequestTimeoutSeconds(int requestTimeoutSeconds) {
+      this.requestTimeoutSeconds = requestTimeoutSeconds;
+      return this;
+    }
+
+    /**
+     * Builds and returns a MomentoBatchUtils instance.
+     *
+     * @return A new instance of MomentoBatchUtils.
+     */
+    public MomentoBatchUtils build() {
+      return new MomentoBatchUtils(cacheClient, maxConcurrentRequests, requestTimeoutSeconds);
+    }
+  }
+
+  public static MomentoBatchUtilsBuilder builder(final CacheClient cacheClient) {
+    return new MomentoBatchUtilsBuilder(cacheClient);
+  }
+
+  /**
+   * Performs a batch get operation for String keys.
+   *
+   * @param cacheName The name of the cache.
+   * @param request The batch get request with String keys.
+   * @return BatchGetResponse The batch get response
+   */
+  public BatchGetResponse batchGet(
+      final String cacheName, final BatchGetRequest.StringKeyBatchGetRequest request) {
+
+    final List<BatchGetResponse.StringKeyBatchGetSummary.GetSummary> summaries = new ArrayList<>();
+
+    for (final String key : request.getKeys()) {
+      final CompletableFuture<GetResponse> getResponseFuture;
+      try {
+        getResponseFuture =
+            executorService
+                .submit(() -> cacheClient.get(cacheName, key))
+                .get(this.requestTimeoutSeconds, TimeUnit.SECONDS);
```

Comment:
This threadpool and the corresponding `get` call is blocking on the _submission_ of the get request, rather than the get request to the server. In theory this will block but not on any network or disk; the `cacheClient.get` call also returns a `Future`, which is returned to the client within the response, and doing a `get()` or `join()` on that will actually block until the server responds back with a result. That is something clients can choose to parallelize, or sequentially process. 

In hindsight, the async nature of our APIs made me choose the decision of wrapping futures within the response, rather than us blocking on them. This threadpool simply makes sure that we _submit_ requests upto `maxConn` in parallel.

Reply:
So it sounds like the calling thread blocks until the cache completable future is initiated, but not necessarily completed. The client consuming the summaries does so when calling join. But doesn't this mean there will be more thanax concurrent requests in flight?

Ie the code will initiate at most that many requests concurrently but you can have more in flight?

As an alternative I expected the get call as part of the lambda passed to the executor. That way there would be at most max concurrent requests in flight at a time.

I realize this is subtle and may be easier to explain synchronously so let me know :-)

Reply:
> But doesn't this mean there will be more thanax concurrent requests in flight?

Yes, there could be. It would have been much easier to work around this if the SDK APIs were sync. One of the other ways we can do this to be accurate is to use the `ScsDataClient` here directly than using the SDK APIs as the former is sync. 

>  realize this is subtle and may be easier to explain synchronously so let me know 

Yes absolutely, will ping you in Seattle AM time to go over this.